File: mybin/se/sysconfig/templates/ItemTemplates/C++/DesignPatterns/Composite/Composite.cpp
#include "$fileinputname$.h"

/**
 * Example of a member function that may be performed on 
 * this object. 
 * TODO: Rename and implement this function
 */
void $safeitemname$Composite::MemberFunction() {
	$safeitemname$Iter it;
	// iterate over each child
	for(it = _children.begin(); it != _children.end(); ++it)
	{
		// call that child's member function
		$safeitemname$Base* child = *it;
		child->MemberFunction();
	}
}

/**
 * Example number property getter.  Simply gets the number
 * property from each child object and adds them.
 * 
 * @return   The sum of all numbers contained in the leaves
 *         under this node in the composite tree.
 */
int $safeitemname$Composite::get_NumberProperty() {
	int sum = 0;
	$safeitemname$Iter it;
	// traverse the children
	for(it = _children.begin(); it != _children.end(); ++it)
	{
		// sum their values
		$safeitemname$Base* child = *it;
		sum += child->get_NumberProperty();
	}
	return sum;
}

/**
 * Example of a numeric property setter
 */
void $safeitemname$Composite::set_NumberProperty(int value) {
	// No-op
}

/**
 * Returns the number of children contained by this composite node
 */
int $safeitemname$Composite::get_ChildCount() {
	return (int)_children.size();
}

/**
 * Returns an iterator for the child collection
 */
$safeitemname$Iter $safeitemname$Composite::Children() {
	return _children.begin();
}

/**
 * Adds a new child to the node
 * 
 * @param child A new child to be added to the child collection
 * 
 * @return The number of children owned by this object
 */
int $safeitemname$Composite::AddChild($safeitemname$Base* child) {
	_children.push_back(child);
	return (int)_children.size();
}

/**
 * Removes a specific child from the node
 * 
 * @param child The child object to be removed from the child
 *              collection
 * 
 * @return The number of children owned by this object
 */
int $safeitemname$Composite::RemoveChild($safeitemname$Base* child) {
	_children.remove(child);
	return (int)_children.size();
}

File: mybin/se/sysconfig/templates/ItemTemplates/Java/DesignPatterns/Composite/Composite.java
import java.util.*;

/**
 * This class provides a standard implementation of the Composite
 * class as part of the Composite design pattern.
 * 
 * @author $username$
 */
public class $safeitemname$Composite extends $safeitemname$Component {

    /**
     * Constructor
     * 
     * @param aNodeName	The name of the node being created
     */
    public $safeitemname$Composite(String aNodeName){
        _nodeName = aNodeName;
    }

    /**
     * Example of a member function that may be performed on either a component
     * or leaf node.
     * TODO: Rename and implement operations in the Component and Leaf classes
     */
    public void operation(){
    }

    /**
     * Dumps the contents of the Composite hierarchy, indenting each level by
     * the amount of space specified in padding.
     * 
     * @param padding A string containing the space to pad each indented level.
     */
    public void dump(String padding) {
        System.out.println(padding + "Composite node: " + _nodeName);
	// traverse the children
        for (Iterator<$safeitemname$Component> i = _componentList.iterator(); i.hasNext(); ) {
    		// have each child dump their contents
                i.next().dump(padding + padding);
            }
    }



}
